src/System.Runtime/src/System/HashCode.cs

```diff
@@ -0,0 +1,428 @@
+            var hc2 = (uint)(value2?.GetHashCode() ?? 0);
+            var hc3 = (uint)(value3?.GetHashCode() ?? 0);
+            var hc4 = (uint)(value4?.GetHashCode() ?? 0);
+            var hc5 = (uint)(value5?.GetHashCode() ?? 0);
+            var hc6 = (uint)(value6?.GetHashCode() ?? 0);
+            var hc7 = (uint)(value7?.GetHashCode() ?? 0);
+            var hc8 = (uint)(value8?.GetHashCode() ?? 0);
+
+            Initialize(out uint v1, out uint v2, out uint v3, out uint v4);
+
+            Round(ref v1, hc1);
+            Round(ref v2, hc2);
+            Round(ref v3, hc3);
+            Round(ref v4, hc4);
+
+            Round(ref v1, hc5);
+            Round(ref v2, hc6);
+            Round(ref v3, hc7);
+            Round(ref v4, hc8);
+
+            uint hash = MixState(v1, v2, v3, v4);
+            hash += 32;
+
+            MixFinal(ref hash);
+            return (int)hash;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint Rol(uint value, int count)
+            => (value << count) | (value >> (32 - count));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Initialize(out uint v1, out uint v2, out uint v3, out uint v4)
+        {
+            var seed = s_seed;
+            v1 = seed + Prime1 + Prime2;
+            v2 = seed + Prime2;
+            v3 = seed + 0;
+            v4 = seed - Prime1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Round(ref uint hash, uint input)
+        {
+            hash += input * Prime2;
+            hash = Rol(hash, 13);
+            hash *= Prime1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void QueueRound(ref uint hash, uint queuedValue)
+        {
+            hash += queuedValue * Prime3;
+            hash = Rol(hash, 17) * Prime4;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint MixState(uint v1, uint v2, uint v3, uint v4)
+        {
+            return Rol(v1, 1) + Rol(v2, 7) + Rol(v3, 12) + Rol(v4, 18);
+        }
+
+        private static uint MixEmptyState()
+        {
+            return s_seed + Prime5;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void MixFinal(ref uint hash)
+        {
+            hash ^= hash >> 15;
+            hash *= Prime2;
+            hash ^= hash >> 13;
+            hash *= Prime3;
+            hash ^= hash >> 16;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add<T>(T value)
+        {
+            Add(value?.GetHashCode() ?? 0);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add<T>(T value, IEqualityComparer<T> comparer)
+        {
+            if (comparer is null)
+                throw new ArgumentNullException(nameof(comparer));
```

Comment:
Should this throw? Existing APIs that take `IEqualityComparer<T>` (e.g. [`Dictionary<TKey, TValue>`](https://github.com/dotnet/coreclr/blob/8ef735c71d18d41b9fc4695cd994b4aa6e14b6b3/src/mscorlib/shared/System/Collections/Generic/Dictionary.cs#L75)) use `EqualityComparer<T>.Default` when the comparer is null.

Reply:
Agreed.